src/lib/int.ts

```diff
@@ -301,119 +299,183 @@ export class UInt32 extends CircuitValue {
+  }
+  static fromBigInt(x: bigint) {
+    let xField = x < 0n ? Field((-x).toString()).neg() : Field(x.toString());
+    return Int64.fromFieldUnchecked(xField);
   }
-}
 
-export class Int64 {
-  // In the range [-2^63, 2^63 - 1]
-  @prop value: Field;
+  toString() {
+    let abs = this.magnitude.toString();
+    let sgn = this.sgn.equals(Field.one).toBoolean() || abs === '0' ? '' : '-';
+    return sgn + abs;
+  }
 
-  static check() {
-    throw 'todo: int64 check';
+  isConstant() {
+    return this.magnitude.value.isConstant() && this.sgn.isConstant();
   }
 
-  /*
-  @prop magnitude: UInt64 | null;
-  @prop isPos: Sgn | null;
-  */
+  // --- circuit-compatible operations below ---
+  // the assumption here is that all Int64 values that appear in a circuit are already checked as valid
+  // this is because Circuit.witness calls .check
+  // so we only have to do additional checks if an operation on valid inputs can have an invalid outcome (example: overflow)
 
-  constructor(x: Field) {
-    this.value = x;
+  static check(x: Int64) {
+    UInt64.check(x.magnitude); // |x| < 2^64
+    x.sgn.square().assertEquals(Field.one); // sign(x)^2 === 1
   }
 
-  toString(): string {
-    const s = this.value.toString();
-    const n = BigInt(s);
-    if (n < 1n << 64n) {
-      return s;
-    } else {
-      return '-' + this.value.neg().toString();
-    }
+  static get zero() {
+    return new Int64(UInt64.zero, Field.one);
   }
-
-  static get zero(): Int64 {
-    return new Int64(Field.zero);
+  static get one() {
+    return new Int64(UInt64.one, Field.one);
   }
-
-  static fromUnsigned(x: UInt64): Int64 {
-    return new Int64(x.value);
+  static get minusOne() {
+    return new Int64(UInt64.one, Field.one.neg());
   }
 
-  private static shift(): Field {
-    return Field.fromJSON((1n << 64n).toString()) as Field;
+  toField() {
+    return this.magnitude.value.mul(this.sgn);
   }
 
-  uint64Value(): Field {
-    const n = BigInt(this.value.toString());
-    if (n < 1n << 64n) {
-      return this.value;
-    } else {
-      const x = this.value.add(Int64.shift());
-
-      return x;
-    }
+  static fromField(x: Field): Int64 {
+    let getUnchecked = () => Int64.fromFieldUnchecked(x);
+    // constant case - just return unchecked value
+    if (x.isConstant()) return getUnchecked();
+    // variable case - create a new checked witness and prove consistency with original field
+    let xInt = Circuit.witness(Int64, getUnchecked);
+    xInt.toField().assertEquals(x); // sign(x) * |x| === x
+    return xInt;
   }
 
-  static sizeInFields(): number {
-    return 1;
+  neg() {
+    // doesn't need further check if `this` is valid
+    return new Int64(this.magnitude, this.sgn.neg());
   }
 
-  neg(): Int64 {
-    return new Int64(this.value.neg());
+  add(y: Int64 | number | string | bigint | UInt64 | UInt32) {
+    let y_ = argToInt64(y);
+    return Int64.fromField(this.toField().add(y_.toField()));
```

Comment:
Yeah I think you may have a point.. I might reconsider this when range checks become cheap.

Do you mean shifting with `2^64 + res` and range-checking that? I don't follow how it works with multiplication & negative numbers.

Currently, it's convenient to have Int64 be {magnitude, sign} because that mirrors the Ocaml data structure. If I make it a single Field constrained to 64 bits, I allow a slightly smaller range than the Ocaml type (64 vs 65 bits)

Reply:
yeah `+` not `*` sorry